ipam/store.go

```diff
@@ -17,85 +17,130 @@ package ipam
+	results := make([]Endpoint, 0)
+	tx.Where(&Endpoint{Ip: ip}).Find(&results)
+	if len(results) == 0 {
+		tx.Rollback()
+		return Endpoint{}, common.NewError404("Endpoint", ip)
+	}
+	if len(results) > 1 {
+		// This cannot happen by constraints...
+		tx.Rollback()
+		errMsg := fmt.Sprintf("Expected one result for ip %s, got %v", ip, results)
+		log.Printf(errMsg)
+		return Endpoint{}, common.NewError500(errors.New(errMsg))
+	}
 
-	row := tx.Model(IpamVm{}).Where("host_id = ? AND segment_id = ?", vm.HostId, vm.SegmentId).Select("IFNULL(MAX(seq),-1)+1").Row()
-	row.Scan(&vm.Seq)
-	log.Printf("New sequence is %d\n", vm.Seq)
-
-	// vmSeq is the sequence number of VM in a given host
-	effectiveVmSeq := getEffectiveSeq(vm.Seq, stride)
-	log.Printf("Effective sequence for seq %d (stride %d): %d\n", vm.Seq, stride, effectiveVmSeq)
-	vm.EffectiveSeq = effectiveVmSeq
-	ipamVm := IpamVm{Vm: *vm}
-	tx.NewRecord(ipamVm)
-	tx = tx.Create(&ipamVm)
+	tx = tx.Model(Endpoint{}).Where("ip = ?", ip).Update("in_use", false)
 	err := common.MakeMultiError(tx.GetErrors())
 	if err != nil {
 		tx.Rollback()
+		return Endpoint{}, err
+	}
+	tx.Commit()
+	return results[0], nil
+}
+
+func (ipamStore *ipamStore) addEndpoint(endpoint *Endpoint, upToEndpointIpInt uint64, stride uint) error {
+	var err error
+	tx := ipamStore.DbStore.Db.Begin()
+	hostId := endpoint.HostId
+	endpoint.InUse = true
+	segId := endpoint.SegmentId
+
+	where := "host_id = ? AND segment_id = ? AND in_use = 0"
+	sel := "min(seq)"
+	log.Printf("Calling SELECT %s FROM endpoints WHERE %s;", sel, fmt.Sprintf(strings.Replace(where, "?", "%s", 2), hostId, segId))
+	row := tx.Model(Endpoint{}).Where(where, hostId, segId).Select(sel).Row()
+	seq := sql.NullInt64{}
+	row.Scan(&seq)
+	log.Printf("minseq: %v", seq)
+	if !seq.Valid {
+		// TODO can this be done in a single query?
+		where = "host_id = ? AND segment_id = ? AND in_use = 1"
+		sel = "ifnull(max(seq),-1)+1"
+		log.Printf("Calling SELECT %s FROM endpoints WHERE %s;", sel, fmt.Sprintf(strings.Replace(where, "?", "%s", 2), hostId, segId))
+		row := tx.Model(Endpoint{}).Where(where, hostId, segId).Select(sel).Row()
+		seq = sql.NullInt64{}
+		row.Scan(&seq)
+		log.Printf("maxseq: %v", seq)
+	}
+	endpoint.Seq = uint64(seq.Int64)
+
+	log.Printf("New sequence is %d\n", endpoint.Seq)
+
+	// EndpointSeq is the sequence number of Endpoint in a given host
+	endpoint.EffectiveSeq = getEffectiveSeq(endpoint.Seq, stride)
+	log.Printf("Effective sequence for seq %d (stride %d): %d\n", endpoint.Seq, stride, endpoint.EffectiveSeq)
+
+	ipInt := upToEndpointIpInt | endpoint.EffectiveSeq
+	endpoint.Ip = common.IntToIPv4(ipInt).String()
+	tx.NewRecord(Endpoint{})
+	tx = tx.Create(endpoint)
+	err = common.MakeMultiError(tx.GetErrors())
+	if err != nil {
+		tx.Rollback()
 		return err
 	}
 	tx.Commit()
 	return nil
 }
 
-// getEffectiveSeq gets effective sequence number of a VM
+// getEffectiveSeq gets effective sequence number of a Endpoint
```

Comment:
"an Endpoint"

Also, might be worth explaining why you call it the 'effective' sequence number. Is there an in-effective one? Or a different one that's not ... effective?